zbus/src/address/transport/unixexec.rs

```diff
@@ -0,0 +1,72 @@
+use std::{
+    borrow::BorrowMut, ffi::OsString, fmt::Display, os::unix::ffi::OsStrExt, path::PathBuf,
+    process::Stdio,
+};
+
+use crate::process::Command;
+
+use super::encode_percents;
+
+/// A unixexec domain socket transport in a D-Bus address.
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct UnixExec {
+    pub(super) path: PathBuf,
+    pub(super) arg0: Option<OsString>,
+    pub(super) args: Vec<String>,
+}
+
+impl UnixExec {
+    /// Create a new unixexec transport with the given path and arguments.
+    pub fn new(path: PathBuf, arg0: Option<OsString>, args: Vec<String>) -> Self {
+        Self { path, arg0, args }
+    }
+
+    pub(super) fn from_options(opts: std::collections::HashMap<&str, &str>) -> crate::Result<Self> {
+        let Some(path) = opts.get("path") else {
+            return Err(crate::Error::Address(
+                "unixexec address is missing `path`".to_owned(),
+            ));
+        };
+
+        let arg0 = opts.get("argv0").map(OsString::from);
+
+        let mut args: Vec<String> = Vec::new();
+        let mut arg_index = 1;
+        while let Some(arg) = opts.get(format!("argv{arg_index}").as_str()) {
+            args.push(arg.to_string());
+            arg_index += 1;
+        }
+
+        Ok(Self::new(PathBuf::from(path), arg0, args))
+    }
+
+    pub(super) async fn connect(self) -> crate::Result<crate::connection::socket::Command> {
+        Command::from(self)
+            .stdin(Stdio::piped())
+            .stdout(Stdio::piped())
+            .stderr(Stdio::inherit())
+            .spawn()?
+            .borrow_mut()
+            .try_into()
+    }
+}
+
+impl Display for UnixExec {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.write_str("unixexec:")?;
+        encode_percents(f, self.path.as_os_str().as_bytes())
```

Comment:
this is not idempotent. You will need to refactor the handling of arguments: `Vec<String>` is not appropriate.

@zeenix
It probably makes more sense to rebase on top of https://github.com/dbus2/zbus/pull/989 instead.

Reply:
@elmarco sure but could you do the rebase for @vially ?

Reply:
The code in #989 leave the handling of missing argv0..N to the caller. I think this makes more sense. "no further argvY for Y > X are taken into account." doesn't mean they are not parsed.

Reply:
it's not explicit about duplicate keys either afaict

Reply:
> @elmarco sure but could you do the rebase for @vially ?

@vially what do you prefer?

Reply:
Thanks for the review. I will rebase on top of #989, address the feedback and add some tests then.

Reply:
@elmarco I have rebased on top of #989, please take a look (especially around the changes to `zbus/src/address/transport/unixexec.rs`, as I'm still in the process of learning Rust and I'm not very familiar with the the `Cow` types).

Still planning to add some tests (once I figure out a good way to test the transport).

Reply:
thanks @vially, the ::address modification and usage look good to me!

Reply:
@elmarco I've pushed another revision where the `From<&Unixexec>` implementation has been moved to `::process`.

@zeenix Can you give me some pointers on how to write some tests for this? I've looked a bit into the `e2e.rs` tests but I haven't found any *easy* way to test this (without significant refactorings in that file which seemed messy and unnecessary).

Reply:
@vially Yeah, I don't think you should mess around with `e2e`. I think where you've currently put the connection test (right in the module) is fine actually. Alternatively it can also go into a separate new file/mod under `tests` dir (which as of yesterday, contains more than `e2e`). I'd prefer the latter since it's not a unit test.

As for how/what to test, I think it'd be sufficient to establish a connection with the bus, perhaps through `systemd-stdio-bridge`. I'm not sure if `systmd-stdio-bridge` is installed on the CI but if it's not, I think we can install it [like we install dbus](https://github.com/dbus2/zbus/blob/bfeb096f3ade97d48b04597f478e303a917e9e3f/.github/workflows/rust.yml#L76).

Reply:
> @elmarco I've pushed another revision where the `From<&Unixexec>` implementation has been moved to `::process`.

cool. Perhaps the last rebase didn't go well cause I see commits from main in this PR, while they don't show in @elmarco's PR branch. :thinking:

Reply:
> Thanks for the feedback.

Np.

> I have added a basic `unixexec` connection test case using `systemd-stdio-bridge`.

Awesome!

> I don't have an easy way to run the CI pipeline to see if `systemd-stdio-bridge` is installed, but if the pipeline fails I'll get it fixed.

That's no issue. Github will run it after I approve it (approval is only needed for first time contributors). The issue is that currently this branch is in conflict with main so CI won't even attempt to build it.

> I have tried moving the socket test into the `tests` directory too, but it would require making the `socket::command::connect` module/method public. If you think it's worth it I can fix that in the next revision.

Sure, that can wait.

Reply:
I've just rebased on `main` which fixed the merge conflicts.